transformers/generate/sampler.py

```diff
@@ -0,0 +1,210 @@
+        to be sampled from. Specifically, choose the smallest set such that the
+        sum of its items' probabilities is greater than a number p in [0,1].
+
+        .. Holtzman, Ari, et al. "The curious case of neural text
+           degeneration." arXiv preprint arXiv:1904.09751 (2019).
+        """
+        if self.p > 0:
+            sorted_logits, sorted_indices = torch.sort(logits, descending=True)
+            sorted_probabilities = F.softmax(sorted_logits, dim=-1)
+            cumulative_probabilities = torch.cumsum(sorted_probabilities, dim=-1)
+
+            # Remove tokens with cumulative probability above the threshold,
+            # but keep the first token above the threshold.
+            sorted_indices_to_remove = cumulative_probabilities > self.p
+            sorted_indices_to_remove[..., 1:] = sorted_indices_to_remove[..., :-1].clone()
+            sorted_indices_to_remove[..., 0] = 0
+
+            # scatter sorted tensors to original indexing
+            indices_to_remove = sorted_indices_to_remove.scatter(
+                dim=-1, index=sorted_indices, src=sorted_indices_to_remove
+            )
+            logits[indices_to_remove] = -float("Inf")
+
+        return logits
+
+    def sample_one_token(self, logits):
+        if self.do_apply_temperature:
+            return torch.multinomial(F.softmax(logits, dim=-1), num_samples=1)
+        return torch.argmax(logits, dim=-1).unsqueeze(-1)
+
+
+class SamplerSingleStack(Sampler):
+    """ Generic sampler for single-stack models.
+    """
+
+    def __init__(self, model, config, device):
+        self.model = model
+        super(SamplerSingleStack, self).__init__(config, device)
+
+    def generate_sequence(self, length=1, prompt_ids=[], **model_kwargs):
+        prompt_ids = torch.tensor(
+            prompt_ids, dtype=torch.long, device=self.device
+        ).unsqueeze(0)
+        generated_sequence = prompt_ids
+        with torch.no_grad():
+            for _ in trange(length):
+                outputs = self.model.decode(generated_sequence, **model_kwargs)
+                next_tokens_logits = outputs[0][:, -1, :]
+                next_tokens = self.generate_one_token(
+                    next_tokens_logits, generated_sequence
+                )
+                generated_sequence = torch.cat((generated_sequence, next_tokens), dim=1)
+
+        return generated_sequence.squeeze(0).tolist()
+
+
+class SamplerEncoderDecoder(Sampler):
+    """ Generic sampler for encoder-decoder models.
+    """
+
+    def __init__(self, model, config, device):
+        self.model = model
+        super(SamplerEncoderDecoder, self).__init__(config, device)
+
+    def generate_sequence(self, encoder_input_ids, length=1, prompt_ids=[], **model_kwargs):
+        encoder_kwargs, decoder_kwargs = self.model.prepare_model_kwargs(**model_kwargs)
+        encoder_hidden_states = self.model.encode(encoder_input_ids, **encoder_kwargs)
```

Comment:
this could use a `torch.no_grad()` as well

Reply:
Also I think `self.model.encode` output a tuple. We should have tests on this new class they would have caught this.

Reply:
Yes, to all. I haven't had time to add tests on this class. Will do.